Adding a delivery date field.

xiinweii98 opened the pull request:
Work in progress. Fixing tests.

Fix #16

xiinweii98 committed:
Adding a delivery date field.

Work in progress. Fixing tests.

xiinweii98 committed:
Adding DeliveryDate field.

Mostly followed Tutorial 3 guide.

pPris commented on src/main/java/seedu/address/model/util/SampleDataUtil.java:
```diff
@@ -1,42 +1,49 @@
             new Person(new Name("Alex Yeoh"), new Phone("87438807"), new Email("alexyeoh@example.com"),
                 new Address("Blk 30 Geylang Street 29, #06-40"),
-                getTagSet("friends")),
+                getTagSet("friends"), new DeliveryDate(dateFormatter.format(dateToday.plusDays(3L)))),
```
I like how you used this plusDays() method in your code, didn't realise it exists

pPris commented on src/main/java/seedu/address/model/person/DeliveryDate.java:
```diff
@@ -0,0 +1,82 @@
+ * Represents an Order's delivery datre in the CakeCollate.
+ * Guarantees: immutable; is valid as declared in {@link #isValidDeliveryDate(String)}
+ */
+public class DeliveryDate {
```
Looks well-implemented to me, seems pretty extensible if we need to add more formats or features

pPris commented on src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json:
```diff
@@ -5,42 +5,49 @@
     "address" : "123, Jurong West Ave 6, #08-111",
-    "tagged" : [ "friends" ]
+    "tagged" : [ "friends" ],
+    "deliveryDate": "01/01/2022"
```
Perhaps you could consider using different dates here so there's more variety in the tests?

xiinweii98 replied:
Yup definitely! If I recall correctly Benson's details will be used specifically in some tests. So since you are adding a field as well, take note of those specific test cases that tries to extract Benson's details ya. Probably don't have to do anything about it if you didn't miss anything ~

pPris replied:
Thanks for the heads up :)

pPris commented on src/test/java/seedu/address/model/person/DeliveryDateTest.java:
```diff
@@ -0,0 +1,48 @@
+
+
+
+public class DeliveryDateTest {
```
Looks pretty thorough to me, nice

pPris replied:
I'm not sure if this may be necessary, but I think it could be good to have two date strings, parse them, and make sure they're equal in the date value represent? Just to check that the equals method doesn't break

xiinweii98 replied:
Good catch! Will do.

pPris commented on src/test/java/seedu/address/model/person/DeliveryDateTest.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Test
+    public void isValidDeliveryDate() {
```
A lot of formats seem to be covered here, nice 👍

pPris commented on src/test/java/seedu/address/model/person/DeliveryDateTest.java:
```diff
@@ -0,0 +1,48 @@
+        assertFalse(DeliveryDate.isValidDeliveryDate("32/01/2022")); // invalid day
+        assertFalse(DeliveryDate.isValidDeliveryDate("01/01/202020")); // invalid year
+        // less than 3 working days
+        assertFalse(DeliveryDate.isValidDeliveryDate(LocalDate.now().plusDays(2L).toString()));
```
Are there any delimiters your delivery date class is intentionally not meant to handle? If so perhaps you can add an invalid delimiter test here

xiinweii98 replied:
Yup, there are. Nice catch again:)

pPris reviewed (commented):
Overall the changes made really thorough and extensible :) I left some suggestions you might want to consider. Also leaving it open so someone else who isn't doing a similar issue can check.

xiinweii98 committed:
Updated according to reviews from PR.

-Edited TypicalPersons and typicalPersonsAddressBook to include variety of date formats
-Edited DeliveryDateTest to test the .equals method
-Edited DeliveryDateTest to test for invalid delimiters and formats

Noteworthy problems that may face in the future:
DeliveryDates of persons/typical persons created in Test Cases are mostly hard-coded, as such, the test cases are not future proof since "Today's" date will keep moving forward.
This is because isValidDeliveryDate checks and makes sure deliveryDate is at least 3 days after "Today's" date.
While it is possible to code .java test files to make use of LocalDate.now() to input DeliveryDates for the test cases, I am currently uncertain if I can do the same for .json test files.
This is important as there are some .json test files that are hard-coded and used for Test Cases and so have a depency.
For now, all valid DeliveryDates used for testing is set for 2022.

xiinweii98 committed:
Fix checkstyleTest errors.

xiinweii98 committed:
Updated "the CakeCollate" to "CakeCollate".
